cmd/calyptia/get_cluster_objects.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"text/tabwriter"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/sync/errgroup"
+	"gopkg.in/yaml.v2"
+
+	cloud "github.com/calyptia/api/types"
+)
+
+func (config *config) fetchAllClusterObjects() ([]cloud.ClusterObject, error) {
+	aa, err := config.cloud.CoreInstances(config.ctx, config.projectID, cloud.CoreInstancesParams{})
+	if err != nil {
+		return nil, fmt.Errorf("could not prefetch core-instances: %w", err)
+	}
+
+	if len(aa.Items) == 0 {
+		return nil, nil
+	}
+
+	var clusterobjects []cloud.ClusterObject
+	var mu sync.Mutex
+
+	g, gctx := errgroup.WithContext(config.ctx)
+	for _, a := range aa.Items {
+		a := a
+		g.Go(func() error {
+			objects, err := config.cloud.ClusterObjects(gctx, a.ID,
+				cloud.ClusterObjectParams{})
+			if err != nil {
+				return err
+			}
+
+			mu.Lock()
+			clusterobjects = append(clusterobjects, objects.Items...)
+			mu.Unlock()
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	var uniqueClusterObjects []cloud.ClusterObject
+	clusterObjectIDs := map[string]struct{}{}
+	for _, coid := range clusterobjects {
+		if _, ok := clusterObjectIDs[coid.ID]; !ok {
+			uniqueClusterObjects = append(uniqueClusterObjects, coid)
+			clusterObjectIDs[coid.ID] = struct{}{}
+		}
+	}
+
+	return uniqueClusterObjects, nil
+}
+
+func (config *config) completeClusterObjects(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	pp, err := config.fetchAllClusterObjects()
+	if err != nil {
+		cobra.CompError(err.Error())
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	if pp == nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	return clusterObjectsKeys(pp), cobra.ShellCompDirectiveNoFileComp
+}
+
+// ClusterObjectsKeys returns unique cluster object names first and then IDs.
+func clusterObjectsKeys(aa []cloud.ClusterObject) []string {
+	namesCount := map[string]int{}
+	for _, a := range aa {
+		if _, ok := namesCount[a.Name]; ok {
+			namesCount[a.Name] += 1
+			continue
+		}
+
+		namesCount[a.Name] = 1
+	}
+
+	var out []string
+
+	for _, a := range aa {
+		var nameIsUnique bool
+		for name, count := range namesCount {
+			if a.Name == name && count == 1 {
+				nameIsUnique = true
+				break
+			}
+		}
+		if nameIsUnique {
+			out = append(out, a.Name)
+			continue
+		}
+
+		out = append(out, a.ID)
+	}
+
+	return out
+}
+
+func (config *config) loadClusterObjectID(key string, environmentID string) (string, error) {
+	aa, err := config.fetchAllClusterObjects()
+	if err != nil {
+		return "", err
+	}
+
+	for _, a := range aa {
+		if a.Name == key {
```

Comment:
This is not the correct way to do it.
There could be more than one cluster object with the same name on the project; considering the name is only unique per core instance.

Take a look at `loadPipelineID` for example.

Reply:
Is there an API to retrieve the ID of a cluster object by name?

Right now I am working on modifying clusterObjectKeys to return all unique names with their IDs.